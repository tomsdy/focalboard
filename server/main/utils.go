package main

import (
	"crypto/rand"
	"fmt"
	"log"
	"os"
)

// FileExists returns true if a file exists at the path
func fileExists(path string) bool {
	_, err := os.Stat(path)
	if os.IsNotExist(err) {
		return false
	}
	return err == nil
}

// CreateGUID returns a random GUID
func createGUID() string {
	b := make([]byte, 16)
	_, err := rand.Read(b)
	if err != nil {
		log.Fatal(err)
	}
	uuid := fmt.Sprintf("%x-%x-%x-%x-%x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:])

	return uuid
}
