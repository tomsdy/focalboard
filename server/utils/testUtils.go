package utils

import "github.com/stretchr/testify/mock"

var Anything = mock.MatchedBy(func(interface{}) bool { return true })
