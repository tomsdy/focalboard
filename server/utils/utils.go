package utils

import (
	"crypto/rand"
	"fmt"
	"log"
	"time"
)

// CreateGUID returns a random GUID.
func CreateGUID() string {
	b := make([]byte, 16)
	_, err := rand.Read(b)
	if err != nil {
		log.Fatal(err)
	}
	uuid := fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])

	return uuid
}

// GetMillis is a convenience method to get milliseconds since epoch.
func GetMillis() int64 {
	return time.Now().UnixNano() / int64(time.Millisecond)
}
