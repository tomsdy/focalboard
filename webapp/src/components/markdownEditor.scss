/* Markdown Editor */

.MarkdownEditor {
	cursor: text;
    /* CodeMirror / SimpleMDE / EasyMDE overrides */

    .CodeMirror {
        padding: 0;
        min-height: 0;
        background: var(--main-bg);
        color: var(--body-color);
    }

    pre.CodeMirror-line {
        padding: 0;
    }

    .CodeMirror, .CodeMirror-scroll {
        flex: 1;
        min-height: 0;
        line-height: 1.3;
        border: none;
    }

    .CodeMirror-cursor {
        border-left: 2px solid var(--cursor-color);
    }

    .CodeMirror-selected {
        background: rgba(147, 204, 231, 0.8) !important;
    }

    .CodeMirror-scroll {
        overflow: hidden !important;
    }
    /* hide the scrollbars */
    .CodeMirror-vscrollbar, .octo-editor .CodeMirror-hscrollbar {
        display: none !important;
    }

    .octo-editor-preview {
        min-height: 30px;
    }

    /* .dialog .octo-editor-preview {
        min-height: 100px;
    } */

    .octo-editor-activeEditor {
        overflow: hidden;
        border: solid 1px #aaccff;
        border-radius: 5px;
    }
}
