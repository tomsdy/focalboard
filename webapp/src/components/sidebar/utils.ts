// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {CategoryBoards, DefaultCategory} from '../../store/sidebar'

import {Block} from '../../blocks/block'
import {Board} from '../../blocks/board'

export function addMissingItems(sidebarCategories: Array<CategoryBoards>, allItems: Array<Block | Board>): Array<CategoryBoards> {
    const blocksInCategories = new Map<string, boolean>()
    sidebarCategories.forEach(
        (category) => category.boardIDs.forEach(
            (boardID) => blocksInCategories.set(boardID, true),
        ),
    )

    const defaultCategory: CategoryBoards = {
        ...DefaultCategory,
        boardIDs: [],
    }

    allItems.forEach((block) => {
        if (!blocksInCategories.get(block.id)) {
            defaultCategory.boardIDs.push(block.id)
        }
    })

    // sidebarCategories comes from store,
    // so is frozen and can't be extended.
    // So creating new array from it.
    return [
        ...sidebarCategories,
        defaultCategory,
    ]
}
