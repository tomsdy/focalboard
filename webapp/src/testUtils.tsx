// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'

export const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()

        range.getBoundingClientRect = jest.fn()

        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }

        return range
    }
}
