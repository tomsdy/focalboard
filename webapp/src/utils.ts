// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import marked from 'marked'

declare global {
    interface Window {
        msCrypto: Crypto
    }
}

class Utils {
    static createGuid(): string {
        const crypto = window.crypto || window.msCrypto
        function randomDigit() {
            if (crypto && crypto.getRandomValues) {
                const rands = new Uint8Array(1)
                crypto.getRandomValues(rands)
                return (rands[0] % 16).toString(16)
            }

            return (Math.floor((Math.random() * 16))).toString(16)
        }
        return 'xxxxxxxx-xxxx-4xxx-8xxx-xxxxxxxxxxxx'.replace(/x/g, randomDigit)
    }

    static htmlToElement(html: string): HTMLElement {
        const template = document.createElement('template')
        template.innerHTML = html.trim()
        return template.content.firstChild as HTMLElement
    }

    static getElementById(elementId: string): HTMLElement {
        const element = document.getElementById(elementId)
        Utils.assert(element, `getElementById "${elementId}$`)
        return element!
    }

    static htmlEncode(text: string): string {
        return String(text).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')
    }

    static htmlDecode(text: string): string {
        return String(text).replace(/&amp;/g, '&').replace(/&lt;/g, '<').replace(/&gt;/g, '>').replace(/&quot;/g, '"')
    }

    // Markdown

    static htmlFromMarkdown(text: string): string {
        // HACKHACK: Somehow, marked doesn't encode angle brackets
        const renderer = new marked.Renderer()
        renderer.link = (href, title, contents) => `<a target="_blank" rel="noreferrer" href="${href}" title="${title || ''}" onclick="event.stopPropagation();">${contents}</a>`
        const html = marked(text.replace(/</g, '&lt;'), {renderer, breaks: true})
        return html
    }

    // Date and Time

    static displayDate(date: Date): string {
        const dateTimeFormat = new Intl.DateTimeFormat('en', {year: 'numeric', month: 'short', day: '2-digit'})
        const text = dateTimeFormat.format(date)

        return text
    }

    static displayDateTime(date: Date): string {
        const dateTimeFormat = new Intl.DateTimeFormat(
            'en',
            {
                year: 'numeric',
                month: 'short',
                day: '2-digit',
                hour: 'numeric',
                minute: 'numeric',
            })
        const text = dateTimeFormat.format(date)
        return text
    }

    static sleep(miliseconds: number): Promise<void> {
        return new Promise((resolve) => setTimeout(resolve, miliseconds))
    }

    // Errors

    static assertValue(valueObject: any): void {
        const name = Object.keys(valueObject)[0]
        const value = valueObject[name]
        if (!value) {
            Utils.logError(`ASSERT VALUE [${name}]`)
        }
    }

    static assert(condition: any, tag = ''): void {
        /// #!if ENV !== "production"
        if (!condition) {
            Utils.logError(`ASSERT [${tag ?? new Error().stack}]`)
        }

        /// #!endif
    }

    static assertFailure(tag = ''): void {
        /// #!if ENV !== "production"
        Utils.assert(false, tag)

        /// #!endif
    }

    static log(message: string): void {
        /// #!if ENV !== "production"
        const timestamp = (Date.now() / 1000).toFixed(2)
        // eslint-disable-next-line no-console
        console.log(`[${timestamp}] ${message}`)

        /// #!endif
    }

    static logError(message: string): void {
        /// #!if ENV !== "production"
        const timestamp = (Date.now() / 1000).toFixed(2)
        // eslint-disable-next-line no-console
        console.error(`[${timestamp}] ${message}`)

        /// #!endif
    }

    // favicon

    static setFavicon(icon?: string): void {
        const href = icon ? `data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100"><text y=".9em" font-size="90">${icon}</text></svg>` : ''
        const link = (document.querySelector("link[rel*='icon']") || document.createElement('link')) as HTMLLinkElement
        link.type = 'image/x-icon'
        link.rel = 'shortcut icon'
        link.href = href
        document.getElementsByTagName('head')[0].appendChild(link)
    }

    // URL

    static replaceUrlQueryParam(paramName: string, value?: string): void {
        const queryString = new URLSearchParams(window.location.search)
        const currentValue = queryString.get(paramName) || ''
        if (currentValue !== value) {
            const newUrl = new URL(window.location.toString())
            if (value) {
                newUrl.searchParams.set(paramName, value)
            } else {
                newUrl.searchParams.delete(paramName)
            }
            window.history.pushState({}, document.title, newUrl.toString())
        }
    }

    // File names

    static sanitizeFilename(filename: string): string {
        // TODO: Use an industrial-strength sanitizer
        let sanitizedFilename = filename
        const illegalCharacters = ['\\', '/', '?', ':', '<', '>', '*', '|', '"', '.']
        illegalCharacters.forEach((character) => {
            sanitizedFilename = sanitizedFilename.replace(character, '')
        })
        return sanitizedFilename
    }

    // File picker

    static selectLocalFile(onSelect?: (file: File) => void, accept = '.jpg,.jpeg,.png'): void {
        const input = document.createElement('input')
        input.type = 'file'
        input.accept = accept
        input.onchange = async () => {
            const file = input.files![0]
            onSelect?.(file)
        }

        input.style.display = 'none'
        document.body.appendChild(input)
        input.click()

        // TODO: Remove or reuse input
    }

    // Arrays

    static arraysEqual(a: readonly any[], b: readonly any[]): boolean {
        if (a === b) {
            return true
        }
        if (a === null || b === null) {
            return false
        }
        if (a === undefined || b === undefined) {
            return false
        }
        if (a.length !== b.length) {
            return false
        }

        for (let i = 0; i < a.length; ++i) {
            if (a[i] !== b[i]) {
                return false
            }
        }
        return true
    }

    static arrayMove(arr: any[], srcIndex: number, destIndex: number): void {
        arr.splice(destIndex, 0, arr.splice(srcIndex, 1)[0])
    }

    // Clipboard

    static copyTextToClipboard(text: string): boolean {
        const textField = document.createElement('textarea')
        textField.innerText = text
        textField.style.position = 'fixed'
        textField.style.opacity = '0'

        document.body.appendChild(textField)
        textField.select()

        let result = false
        try {
            result = document.execCommand('copy')
        } catch (err) {
            Utils.logError(`copyTextToClipboard ERROR: ${err}`)
            result = false
        }
        textField.remove()

        return result
    }

    static isMobile() {
        const toMatch = [
            /Android/i,
            /webOS/i,
            /iPhone/i,
            /iPad/i,
            /iPod/i,
            /BlackBerry/i,
            /Windows Phone/i,
        ]

        return toMatch.some((toMatchItem) => {
            return navigator.userAgent.match(toMatchItem)
        })
    }
}

export {Utils}
