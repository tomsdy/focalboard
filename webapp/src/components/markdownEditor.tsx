// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import EasyMDE from 'easymde'
import React from 'react'
import SimpleMDE from 'react-simplemde-editor'

import {Utils} from '../utils'
import './markdownEditor.scss'

type Props = {
    text?: string
    placeholderText?: string
    uniqueId?: string
    className?: string
    readonly?: boolean

    onChange?: (text: string) => void
    onFocus?: () => void
    onBlur?: (text: string) => void
    onAccept?: (text: string) => void
}

type State = {
    isEditing: boolean
}

class MarkdownEditor extends React.Component<Props, State> {
    static defaultProps = {
        text: '',
    }

    get text(): string {
        return this.elementRef.current!.state.value || ''
    }
    set text(value: string) {
        this.elementRef.current!.setState({value})
    }

    private editorInstance?: EasyMDE
    private frameRef = React.createRef<HTMLDivElement>()
    private elementRef = React.createRef<SimpleMDE>()
    private previewRef = React.createRef<HTMLDivElement>()

    constructor(props: Props) {
        super(props)
        this.state = {isEditing: false}
    }

    shouldComponentUpdate(): boolean {
        return true
    }

    componentDidUpdate(): void {
        const newText = this.props.text || ''
        if (!this.state.isEditing && this.text !== newText) {
            this.text = newText
        }
    }

    showEditor(): void {
        const cm = this.editorInstance?.codemirror
        if (cm) {
            setTimeout(() => {
                cm.refresh()
                cm.focus()
                cm.getInputField()?.focus()
                cm.setCursor(cm.lineCount(), 0) // Put cursor at end
            }, 100)
        }

        this.setState({isEditing: true})
    }

    hideEditor(): void {
        this.editorInstance?.codemirror?.getInputField()?.blur()
        this.setState({isEditing: false})
    }

    render(): JSX.Element {
        const {text, placeholderText, uniqueId, onFocus, onBlur, onChange} = this.props

        let html: string
        if (text) {
            html = Utils.htmlFromMarkdown(text)
        } else {
            html = Utils.htmlFromMarkdown(placeholderText || '')
        }

        const previewElement = (
            <div
                ref={this.previewRef}
                className={text ? 'octo-editor-preview' : 'octo-editor-preview octo-placeholder'}
                style={{display: this.state.isEditing ? 'none' : undefined}}
                dangerouslySetInnerHTML={{__html: html}}
                onClick={() => {
                    if (!this.props.readonly && !this.state.isEditing) {
                        this.showEditor()
                    }
                }}
            />)

        const editorElement = (
            <div
                className='octo-editor-activeEditor'

                // Use visibility instead of display here so the editor is pre-rendered, avoiding a flash on showEditor
                style={this.state.isEditing ? {} : {visibility: 'hidden', position: 'absolute', top: 0, left: 0}}
                onKeyDown={(e) => {
                    // HACKHACK: Need to handle here instad of in CodeMirror because that breaks auto-lists
                    if (e.keyCode === 27 && !e.shiftKey && !(e.ctrlKey || e.metaKey) && !e.altKey) { // Esc
                        this.editorInstance?.codemirror?.getInputField()?.blur()
                    } else if (e.keyCode === 13 && !e.shiftKey && (e.ctrlKey || e.metaKey) && !e.altKey) { // Cmd+Enter
                        this.editorInstance?.codemirror?.getInputField()?.blur()

                        // HACKHACK: Call onAccept after visual state change
                        setTimeout(() => {
                            Utils.log('onAccept')
                            this.props.onAccept?.(this.elementRef.current!.state.value)
                        }, 20)
                    }
                }}
            >
                <SimpleMDE
                    id={uniqueId}
                    ref={this.elementRef}
                    getMdeInstance={(instance) => {
                        this.editorInstance = instance

                        // BUGBUG: This breaks auto-lists
                        // instance.codemirror.setOption("extraKeys", {
                        //     "Ctrl-Enter": (cm) => {
                        //         cm.getInputField().blur()
                        //     }
                        // })
                    }}
                    value={text}

                    events={{
                        change: () => {
                            if (this.state.isEditing) {
                                const newText = this.elementRef.current!.state.value
                                onChange?.(newText)
                            }
                        },
                        blur: () => {
                            const newText = this.elementRef.current!.state.value
                            const oldText = this.props.text || ''
                            if (newText !== oldText && onChange) {
                                const newHtml = newText ? Utils.htmlFromMarkdown(newText) : Utils.htmlFromMarkdown(placeholderText || '')
                                this.previewRef.current!.innerHTML = newHtml
                                onChange(newText)
                            }

                            this.text = newText

                            this.frameRef.current!.classList.remove('active')

                            if (onBlur) {
                                onBlur(newText)
                            }

                            this.hideEditor()
                        },
                        focus: () => {
                            this.frameRef.current?.classList.add('active')
                            this.elementRef.current?.setState({value: this.text})

                            if (onFocus) {
                                onFocus()
                            }
                        },
                    }}
                    options={{
                        autoDownloadFontAwesome: true,
                        toolbar: false,
                        status: false,
                        spellChecker: false,
                        minHeight: '10px',
                        shortcuts: {
                            toggleStrikethrough: 'Cmd-.',
                            togglePreview: null,
                            drawImage: null,
                            drawLink: null,
                            toggleSideBySide: null,
                            toggleFullScreen: null,
                        },
                    }}
                />
            </div>)

        const element = (
            <div
                ref={this.frameRef}
                className={`MarkdownEditor octo-editor ${this.props.className || ''}`}
            >
                {previewElement}
                {editorElement}
            </div>)

        return element
    }
}

export {MarkdownEditor}
