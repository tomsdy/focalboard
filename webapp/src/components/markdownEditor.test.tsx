// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {act, fireEvent, render, screen} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

import {mocked} from 'ts-jest/utils'

import {mockDOM, wrapDNDIntl} from '../testUtils'

import {Utils} from '../utils'

import {MarkdownEditor} from './markdownEditor'

jest.mock('../utils')
const mockedUtils = mocked(Utils, true)
jest.useFakeTimers()

describe('components/markdownEditor', () => {
    beforeAll(mockDOM)
    beforeEach(jest.clearAllMocks)
    test('should match snapshot', async () => {
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <MarkdownEditor
                    id={'test-id'}
                    text={''}
                    placeholderText={'placeholder'}
                    className={'classname-test'}
                    readonly={false}

                    onChange={jest.fn()}
                    onFocus={jest.fn()}

                    onBlur={jest.fn()}
                    onAccept={jest.fn()}
                />,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })
    test('return markdownEditor, write hello and verify blur', async () => {
        const onMockedBlur = jest.fn()
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <MarkdownEditor
                    id={'test-id'}
                    text={''}
                    placeholderText={'placeholder'}
                    className={'classname-test'}
                    readonly={false}

                    onChange={jest.fn()}
                    onFocus={jest.fn()}

                    onBlur={onMockedBlur}
                    onAccept={jest.fn()}
                />,
            ))
            container = result.container
            const elementMarkDown = screen.getByRole('textbox', {hidden: true})
            userEvent.click(elementMarkDown)
        })
        const elementsTextArea = screen.getAllByRole('textbox', {hidden: true})
        expect(elementsTextArea).not.toBeNull()
        expect(elementsTextArea.length).toBeGreaterThanOrEqual(2)
        userEvent.type(elementsTextArea[1], 'hello')
        fireEvent.blur(elementsTextArea[1])
        expect(onMockedBlur).toBeCalledTimes(1)
        expect(onMockedBlur).toBeCalledWith('hello')
        expect(container).toMatchSnapshot()
    })

    test('return markdownEditor, write hi and verify change', async () => {
        const onMockedChange = jest.fn()
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <MarkdownEditor
                    id={'test-id'}
                    text={''}
                    placeholderText={'placeholder'}
                    className={'classname-test'}
                    readonly={false}

                    onChange={onMockedChange}
                    onFocus={jest.fn()}

                    onBlur={jest.fn()}
                    onAccept={jest.fn()}
                />,
            ))
            container = result.container
            const elementMarkDown = screen.getByRole('textbox', {hidden: true})
            userEvent.click(elementMarkDown)
        })
        const elementsTextArea = screen.getAllByRole('textbox', {hidden: true})
        expect(elementsTextArea).not.toBeNull()
        expect(elementsTextArea.length).toBeGreaterThanOrEqual(2)
        const elementText = elementsTextArea[1]
        userEvent.type(elementText, 'h')
        expect(onMockedChange).toBeCalledTimes(1)
        expect(onMockedChange).toBeCalledWith('h')
        userEvent.type(elementText, 'i')
        expect(onMockedChange).toBeCalledTimes(2)
        expect(onMockedChange).toBeCalledWith('hi')
        expect(container).toMatchSnapshot()
    })
    test('return markdownEditor and verify accept', async () => {
        const onMockedAccept = jest.fn()
        await act(async () => {
            render(wrapDNDIntl(
                <MarkdownEditor
                    id={'test-id'}
                    text={''}
                    placeholderText={'placeholder'}
                    className={'classname-test'}
                    readonly={false}

                    onChange={jest.fn()}
                    onFocus={jest.fn()}

                    onBlur={jest.fn()}
                    onAccept={onMockedAccept}
                />,
            ))
            const elementMarkDown = screen.getByRole('textbox', {hidden: true})
            userEvent.click(elementMarkDown)
        })
        const elementsTextArea = screen.getAllByRole('textbox', {hidden: true})
        expect(elementsTextArea).not.toBeNull()
        expect(elementsTextArea.length).toBeGreaterThanOrEqual(2)
        fireEvent.click(elementsTextArea[1])
        fireEvent.keyDown(elementsTextArea[1], {metaKey: true, keyCode: 13})
        act(() => {
            jest.runOnlyPendingTimers()
        })
        expect(mockedUtils.log).toBeCalledTimes(1)
        expect(onMockedAccept).toBeCalledTimes(1)
    })
    test('return markdownEditor and verify blur on escape', async () => {
        const onMockedBlur = jest.fn()
        await act(async () => {
            render(wrapDNDIntl(
                <MarkdownEditor
                    id={'test-id'}
                    text={''}
                    placeholderText={'placeholder'}
                    className={'classname-test'}
                    readonly={false}

                    onChange={jest.fn()}
                    onFocus={jest.fn()}

                    onBlur={onMockedBlur}
                    onAccept={jest.fn()}
                />,
            ))
            const elementMarkDown = screen.getByRole('textbox', {hidden: true})
            userEvent.click(elementMarkDown)
        })
        const elementsTextArea = screen.getAllByRole('textbox', {hidden: true})
        expect(elementsTextArea).not.toBeNull()
        expect(elementsTextArea.length).toBeGreaterThanOrEqual(2)
        fireEvent.click(elementsTextArea[1])
        fireEvent.keyDown(elementsTextArea[1], {keyCode: 27})
        expect(onMockedBlur).toBeCalledTimes(1)
    })
})
