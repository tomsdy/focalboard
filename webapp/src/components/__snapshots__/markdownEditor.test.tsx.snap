// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/markdownEditor return markdownEditor, write hello and verify blur 1`] = `
<div>
  <div
    class="MarkdownEditor octo-editor classname-test "
  >
    <div
      class="octo-editor-preview octo-placeholder"
      style=""
    />
    <div
      class="octo-editor-active Editor"
      style="visibility: hidden; position: absolute; top: 0px; left: 0px;"
    >
      <div
        id="test-id-wrapper"
      >
        <textarea
          id="test-id"
          style="display: none;"
        />
        <div
          class="EasyMDEContainer"
        >
          <div
            class="CodeMirror cm-s-easymde CodeMirror-wrap"
          >
            <div
              style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 0px;"
            >
              <textarea
                autocapitalize="off"
                autocorrect="off"
                spellcheck="false"
                style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
                tabindex="0"
              />
            </div>
            <div
              class="CodeMirror-vscrollbar"
              cm-not-content="true"
              tabindex="-1"
            >
              <div
                style="min-width: 1px;"
              />
            </div>
            <div
              class="CodeMirror-hscrollbar"
              cm-not-content="true"
              tabindex="-1"
            >
              <div
                style="height: 100%; min-height: 1px;"
              />
            </div>
            <div
              class="CodeMirror-scrollbar-filler"
              cm-not-content="true"
            />
            <div
              class="CodeMirror-gutter-filler"
              cm-not-content="true"
            />
            <div
              class="CodeMirror-scroll"
              style="min-height: 10px;"
              tabindex="-1"
            >
              <div
                class="CodeMirror-sizer"
                style="margin-left: 0px;"
              >
                <div
                  style="position: relative;"
                >
                  <div
                    class="CodeMirror-lines"
                    role="presentation"
                  >
                    <div
                      role="presentation"
                      style="position: relative; outline: none;"
                    >
                      <div
                        class="CodeMirror-measure"
                      />
                      <div
                        class="CodeMirror-measure"
                      >
                        <pre
                          class="CodeMirror-line"
                          role="presentation"
                        >
                          <span
                            role="presentation"
                            style="padding-right: .1px;"
                          >
                            hello
                          </span>
                        </pre>
                      </div>
                      <div
                        style="position: relative; z-index: 1;"
                      />
                      <div
                        class="CodeMirror-cursors"
                      />
                      <div
                        class="CodeMirror-code"
                        role="presentation"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                style="position: absolute; height: 50px; width: 1px;"
              />
              <div
                class="CodeMirror-gutters"
                style="display: none;"
              />
            </div>
          </div>
          <div
            class="editor-preview-side editor-preview"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/markdownEditor return markdownEditor, write hi and verify change 1`] = `
<div>
  <div
    class="MarkdownEditor octo-editor classname-test active"
  >
    <div
      class="octo-editor-preview octo-placeholder"
      style="display: none;"
    />
    <div
      class="octo-editor-active Editor"
      style=""
    >
      <div
        id="test-id-wrapper"
      >
        <textarea
          id="test-id"
          style="display: none;"
        />
        <div
          class="EasyMDEContainer"
        >
          <div
            class="CodeMirror cm-s-easymde CodeMirror-wrap CodeMirror-focused"
          >
            <div
              style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 0px; left: 0px;"
            >
              <textarea
                autocapitalize="off"
                autocorrect="off"
                spellcheck="false"
                style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
                tabindex="0"
              />
            </div>
            <div
              class="CodeMirror-vscrollbar"
              cm-not-content="true"
              tabindex="-1"
            >
              <div
                style="min-width: 1px;"
              />
            </div>
            <div
              class="CodeMirror-hscrollbar"
              cm-not-content="true"
              tabindex="-1"
            >
              <div
                style="height: 100%; min-height: 1px;"
              />
            </div>
            <div
              class="CodeMirror-scrollbar-filler"
              cm-not-content="true"
            />
            <div
              class="CodeMirror-gutter-filler"
              cm-not-content="true"
            />
            <div
              class="CodeMirror-scroll"
              style="min-height: 10px;"
              tabindex="-1"
            >
              <div
                class="CodeMirror-sizer"
                style="margin-left: 0px;"
              >
                <div
                  style="position: relative;"
                >
                  <div
                    class="CodeMirror-lines"
                    role="presentation"
                  >
                    <div
                      role="presentation"
                      style="position: relative; outline: none;"
                    >
                      <div
                        class="CodeMirror-measure"
                      />
                      <div
                        class="CodeMirror-measure"
                      >
                        <pre
                          class="CodeMirror-line"
                          role="presentation"
                        >
                          <span
                            role="presentation"
                            style="padding-right: .1px;"
                          >
                            hi
                          </span>
                        </pre>
                      </div>
                      <div
                        style="position: relative; z-index: 1;"
                      />
                      <div
                        class="CodeMirror-cursors"
                      />
                      <div
                        class="CodeMirror-code"
                        role="presentation"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                style="position: absolute; height: 50px; width: 1px;"
              />
              <div
                class="CodeMirror-gutters"
                style="display: none;"
              />
            </div>
          </div>
          <div
            class="editor-preview-side editor-preview"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/markdownEditor should match snapshot 1`] = `
<div>
  <div
    class="MarkdownEditor octo-editor classname-test "
  >
    <div
      class="octo-editor-preview octo-placeholder"
    />
    <div
      class="octo-editor-active Editor"
      style="visibility: hidden; position: absolute; top: 0px; left: 0px;"
    >
      <div
        id="test-id-wrapper"
      >
        <textarea
          id="test-id"
          style="display: none;"
        />
        <div
          class="EasyMDEContainer"
        >
          <div
            class="CodeMirror cm-s-easymde CodeMirror-wrap"
          >
            <div
              style="overflow: hidden; position: relative; width: 3px; height: 0px;"
            >
              <textarea
                autocapitalize="off"
                autocorrect="off"
                spellcheck="false"
                style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"
                tabindex="0"
              />
            </div>
            <div
              class="CodeMirror-vscrollbar"
              cm-not-content="true"
              tabindex="-1"
            >
              <div
                style="min-width: 1px;"
              />
            </div>
            <div
              class="CodeMirror-hscrollbar"
              cm-not-content="true"
              tabindex="-1"
            >
              <div
                style="height: 100%; min-height: 1px;"
              />
            </div>
            <div
              class="CodeMirror-scrollbar-filler"
              cm-not-content="true"
            />
            <div
              class="CodeMirror-gutter-filler"
              cm-not-content="true"
            />
            <div
              class="CodeMirror-scroll"
              style="min-height: 10px;"
              tabindex="-1"
            >
              <div
                class="CodeMirror-sizer"
                style="margin-left: 0px;"
              >
                <div
                  style="position: relative;"
                >
                  <div
                    class="CodeMirror-lines"
                    role="presentation"
                  >
                    <div
                      role="presentation"
                      style="position: relative; outline: none;"
                    >
                      <div
                        class="CodeMirror-measure"
                      >
                        <pre
                          class="CodeMirror-line-like"
                        >
                          <span>
                            xxxxxxxxxx
                          </span>
                        </pre>
                      </div>
                      <div
                        class="CodeMirror-measure"
                      />
                      <div
                        style="position: relative; z-index: 1;"
                      />
                      <div
                        class="CodeMirror-cursors"
                      />
                      <div
                        class="CodeMirror-code"
                        role="presentation"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                style="position: absolute; height: 50px; width: 1px;"
              />
              <div
                class="CodeMirror-gutters"
                style="display: none;"
              />
            </div>
          </div>
          <div
            class="editor-preview-side editor-preview"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
